leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt

```diff
@@ -0,0 +1,368 @@
+  private const val BREAKING_CHARACTER_LENGTH = 1
+  const val BREAK_CHARACTER = '\n'
+
+  /**
+   * This implements a greedy wrapping algorithm.
+   *
+   * <p>
+   *
+   * Each line is split into words, using spaces and dots as separators. From there, a wrapped
+   * version of the line is built by adding one word after the other until the string doesn't fit
+   * anymore within the bounds defined by `width`. After which, a line break is inserted and the process
+   * starts over with what's left of the unwrapped line.
+   *
+   * For each added word, a verification is made to determine whether it was underlined in the original
+   * unwrapped [String]. This is done by comparing the absolute position of the word against the
+   * [MarkerMetadata.position] attribute of the `markerMetadata` parameter. If there is match, a new
+   * line is added after the current one to position [MarkerMetadata.length] amount of '~' at the
+   * [MarkerMetadata.position].
+   *
+   * Finally, all lines start with an offset which includes a decorator character and some level of
+   * indentation.
+   */
+  fun wrap(
+    heapAnalysisStr: String,
+    width: Int
+  ): String {
+    var wrappedTrace = ""
+    val lines = heapAnalysisStr.lines()
+
+    for (l in lines.indices) {
+      val currentLine = lines[l]
+
+      if (currentLine.contains(TILDE_CHARACTER)) {
+        check(l > 0) {
+          "A $TILDE_CHARACTER character cannot be placed on the first line of a leak trace"
+        }
+        // '~' markers that underline references which are on a line that will be wrapped, will be added
+        // manually during the wrapping.
+        if (lines[l - 1].length > width) {
+          continue
+        }
+      }
+      if (currentLine.length <= width) {
+        wrappedTrace = wrappedTrace.plus("${currentLine}$BREAK_CHARACTER")
+        continue
+      }
+
+      val nextWrappedTrace = doWrap(
+          currentLine, findMarkerInfoOn(lines[l + 1]), width
+      )
+      wrappedTrace = wrappedTrace.plus("$nextWrappedTrace$BREAK_CHARACTER")
+    }
+    return wrappedTrace.removeSuffix("$BREAK_CHARACTER")
+  }
+
+  private fun findMarkerInfoOn(line: String): MarkerMetadata? {
+    if (line.contains(TILDE_CHARACTER)) {
+      val lastMarkerAt = line.indexOfLast { it == TILDE_CHARACTER }
+      val firstMarkerAt = line.indexOfFirst { it == TILDE_CHARACTER }
+      return MarkerMetadata(firstMarkerAt, lastMarkerAt - firstMarkerAt + 1)
+    }
+    return null
+  }
+
+  private fun doWrap(
+    line: String,
+    markerMetadata: MarkerMetadata?,
+    width: Int
+  ): String {
+    var offset = ""
+    DecoratedIndentation.get(line)
+        ?.let {
+          offset = "${it.decorationString}${"$SPACE_CHARACTER".repeat(it.totalIndentation)}"
+        }
+
+    var wrappedString = ""
+    var totalWrappedLines = 0
+
+    // used to accurately track the marker position
+    var totalRemovedSpaces = 0
+
+    val stack = Stack<List<String>>()
```

Comment:
The JDK `Stack` class should be avoided. ArrayList and ArrayDeque are good alternatives.

Reply:
The `ArrayDeque` from kotlin-stdlib is still marked as experimental. Are we okay with that? Or do you advise using the JDK one?